server/src/main/java/org/apache/uniffle/server/ShuffleDataFlushEvent.java

```diff
@@ -36,6 +38,9 @@ public class ShuffleDataFlushEvent {
   private final Supplier<Boolean> valid;
   private final ShuffleBuffer shuffleBuffer;
   private final AtomicInteger retryTimes = new AtomicInteger();
+  private boolean isPended = false;
+  private StorageManager underStorageManager;
```

Comment:
Could we avoid using storageManager in a ShuffleDataFlushEvent? It's a little weird.

Reply:
Yes. I also want to avoid this, but if we don't, we have to maintain this relation mapping in multipleStorageManager. However we couldn't remove this structure when event's lifecycle finished, because we don't similar interface in storage manager

Reply:
Could we only have `Storage` field. Could we use `Storage` to find its StorageManager?

Reply:
Currently it can't. But this solution sounds great. We could extend the interface of `storage`

Reply:
> Currently it can't. But this solution sounds great. We could extend the interface of `storage`

I think we need add a interface for StorageManager#isBelongto(Storage storage). Storage is in the module `storage`. StorageManager is in the module server. Storage shouldn't depend on StorageManager.

Reply:
Emm.. But I just want to bound the concrete `storageManager` into event here for writing in `MultiStorageManager`.

Reply:
> I have two directions to improve this part of MultiStorageManager
> 
> 1. Remove the `storageManagerCache` , it looks unused and bring some problems. One possible problem I have seen is that, when one event enters into pending queue due to storage cannot write, it will not have a chance to get a new fallback strategy due to cache.
> 2. Avoid changing the external reference object in `write` method. If we want to fallback write to other storage. We could use the invoking sequence like. (selectStorage -> write)  if failed and then (selectStorage -> write), instead of selectStorage -> write (failed) -> write(choose other storage in this method). That means we should change the fallback strategy invoked from `write` method to `selectStorage` method.

@zuston If we use this solution you proposed in #235, i think we don't need maintain this relation

Reply:
> Emm.. But I just want to bound the concrete `storageManager` into event here for writing in `MultiStorageManager`.

Relation is not good idea. But it's better than putting StorageManager into event. If StorageManager is ok, should we put ShuffleFlushManager into the ShuffleFlushEvent instead of  `Supplier<Boolean> valid`?

Reply:
> Relation is not good idea. But it's better than putting StorageManager into event. If StorageManager is ok, should we put ShuffleFlushManager into the ShuffleFlushEvent instead of Supplier<Boolean> valid?

Sounds reasonable. The only thing is that how to clean up this relation when event is flushed.

> @zuston If we use this solution you proposed in https://github.com/apache/incubator-uniffle/pull/235, i think we don't need maintain this relation

The relation will help event find the concrete under-storage-manager selected by method of `selectStorage` when invoked by `write` method. Right? @xianjingfeng

Reply:
Do we have ideas on this problem?

Reply:
> Relation is not good idea. But it's better than putting StorageManager into event. If StorageManager is ok, should we put ShuffleFlushManager into the ShuffleFlushEvent instead of Supplier<Boolean> valid?

I have removed the `storageManager` params in event and reuse the cache to keep relations. Although this is a little bit weird.

Reply:
Emmm…. I still think cache is not suitable. Because it holds the reference of events and will cause memory leak. We should implicitly remove this event reference when event is flushed.

Reply:
> Emmm…. I still think cache is not suitable. Because it holds the reference of events and will cause memory leak. We should implicitly remove this event reference when event is flushed.

It's not acceptable for me to put StorageManager into ShuffleDataFlushEvent. I would rather give up this strategy.

Reply:
I know. I don’t insist on keeping this into event. I just explain the cache is also unreasonable. 

And this is also a potential bug due to memory leak.

Reply:
Could we add a callback for ShuffleFlushEvent. In the callback, it will remove the data. When we call the     cleanupFlushEventData(event) method. We call event.clearCallback();

Reply:
Sounds great!

Reply:
> The relation will help event find the concrete under-storage-manager selected by method of `selectStorage` when invoked by `write` method. Right? @xianjingfeng

Yes, if we remove pending queue, we don't need maintain this relation.